devdoc/cancellation_token_requirements.md

```diff
@@ -0,0 +1,138 @@
+# Cancellation Token Requirements
+
+## Overview
+
+The cancellation token module implements an object that can be used to propagate cancellation signals through multi-threaded programs. The design is roughly equivalent to the [`CancellationToken`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.cancellationtoken?view=net-8.0) type in .NET. The module allows applications to create cancellation token handles which can then be passed around to various functions. The following operations can be performed on cancellation tokens:
```

Comment:
>The design is roughly equivalent to the [`CancellationToken`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.cancellationtoken?view=net-8.0) type in .NET. [](http://example.com/codeflow?start=135&length=169)

I know you say "roughly" but there is one point that may be interesting. The .NET `Register` function says: "If this token is already in the canceled state, the delegate will be run immediately and synchronously." It looks like we differ there.

Reply:
Good catch! Perhaps we can do the same thing.

Reply:
Added requirement `SRS_CANCELLATION_TOKEN_04_022` and implemented it:

> If the cancellation token is already in canceled state, then `cancellation_token_register_notify` shall immediately call `on_cancel` and take no further action and return `NULL`.